index.rst

```diff
@@ -447,6 +447,64 @@ in some contexts where materializing the set would be expensive. The
 'DataRefSet' is usually generated by listing existing datasets of a particular
 dataset type, but its component 'DataRef's can be used with other dataset types.
 
+Caching
+-------
+
+When an object is read by the Butler, the Butler keeps a weakref to that object
+and returns a normal reference to that object. As long as the normal ref is kept
+(i.e. as long as the object is not garbage collected) the Butler will keep the
+weakref. If that object is read again, the butler will get a normal ref (via the
+weakref) from the cache instead of reading the object from persisted data a
+second time.
+
+This is accomplished by keeping a WeakValueDictionary in Butler, where the key
+is a custom hash of the ButlerLocation passed to Butler._read, and the value is
+the object that was loaded using that ButlerLocation. The hash of ButlerLocation
+parameters includes the items needed to uniquely identify an object to be loaded
+(or retrieved from the cache).
+
+pythonType
+    The type of python object that should be returned.
+
+cppType
+    The type of swigged cpp object that will get read & returned (if any)
+
+storageName
+    The name/type of storage that the object is persisted as
+
+id(mapper)
+    Identifier for exactly which mapper instance mapped this object.
+
+id(storage)
+    Identifier for exactly which storage instance read this object.
+
+locationList
+    The URI that the object was read from (I think the length of this list will
+    always be exactly 1.
+
+json dump of usedDataId
+    The usedDataId contains the dataId that was used to map the object. It
+    containes items from the dataId that was passed to butler.get that were used
+    (and not items that were not used), as well as items that had to be looked
+    up.
+
+Cached-Object Mutability
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+This introduces an issue with object mutability. If a component object in the
+cache is shared between two composite objects, then it is possible that one of
+the composite objects could change the component object where the other
+composite object did not expect the component to change.
+
+One way to solve this would be to make composite objects that can/should be
+shared const, but that's c++ talk; it is not obvious that there is any 'good'
+way to make a Python object immutable.
+
+We might need to add api to butler.get so that an object can (should) declare if
+an object will be mutates (in which case the loaded object should be unique.
+This might be difficult in the presence of component objects, however. The 'will
+mutate' flag might need to be in the datasetType definition.
+
```

Comment:
I don't think having a `will mutate` flag in the `datasetType` definition would be any help.  The `datasetTypes` that we'll want to occasionally mutate are the same ones we'll usually want cached.